### PR TITLE
Support wildcards for coverage report paths

### LINKED ISSUE #578: Support wildcards for coverage report paths
Currently property for coverage LCOV report paths doesn't accept paths with wildcards. This might be very useful when path is not stable (e.g. see [question on SOF](http://stackoverflow.com/questions/43319073/wildcards-for-lcov-info-path-in-the-sonarqube-scanners)).

Property name: **sonar.javascript.lcov.reportPaths**

* [Docs for coverage import in SonarJS](https://docs.sonarqube.org/display/PLUG/JavaScript+Coverage+Results+Import)
* [Docs for patterns in SonarQube](https://docs.sonarqube.org/display/SONAR/Narrowing+the+Focus#NarrowingtheFocus-patterns)

We should decide whether this feature should be implemented for old properties (still supported for SQ < 6.2):
- **sonar.javascript.lcov.reportPath**
- **sonar.javascript.lcov.itReportPath**

**Example:**
_"sonar.javascript.lcov.reportPaths=reports/*"_ would match 
_reports/report1.lcov
reports/report2.lcov_

_"sonar.javascript.lcov.reportPaths=**/report.lcov"_ would match 
_reports1/report.lcov
reports2/report.lcov_

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #578

### CODE REVIEW

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.io.filefilter.IOFileFilter;
+import org.apache.commons.io.filefilter.TrueFileFilter;
+import org.sonar.api.utils.WildcardPattern;
+
+// Taken from https://github.com/SonarSource/sonar-python/blob/7a683a8da47ed8db51378fbc6c50a76d98b45097/sonar-python-plugin/src/main/java/org/sonar/plugins/python/DirectoryScanner.java
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java:31]: yeah, we should definitely move it into commons

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java [hunk 2]
```diff
@@ -64,18 +63,33 @@ public void execute(SensorContext context) {
...
       return;
     }
     saveCoverageFromLcovFiles(context, lcovFiles);
   }
 
+  private static List<File> getLcovFiles(File baseDir, Set<String> reportPaths) {
+    List<File> lcovFiles = new ArrayList<>();
+    for (String reportPath : reportPaths) {
+      LOG.debug("Using pattern '{}' to resolve LCOV files", reportPath);
+      DirectoryScanner scanner = new DirectoryScanner(baseDir, WildcardPattern.create(reportPath));
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java:78]: how about passing `reportPath` directly to constructor to hide this `WildcardPattern` usage?

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+public class DirectoryScanner {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java:33]: I would call this class `PatternDirectoryScanner`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  private final File baseDir;
+  private final WildcardPattern pattern;
+
+  public DirectoryScanner(File baseDir, WildcardPattern pattern) {
+    this.baseDir = baseDir;
+    this.pattern = pattern;
+  }
+
+  public List<File> getIncludedFiles() {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java:43]: `getMatchingFiles`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java [hunk 5]
```diff
@@ -64,18 +63,33 @@ public void execute(SensorContext context) {
...
+      List<File> includedFiles = scanner.getIncludedFiles();
+      if (includedFiles.isEmpty()) {
+        File file = new File(reportPath);
+        if (!file.exists()) {
+          LOG.debug("No LCOV files were found using pattern {}", reportPath);
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java:83]: I think it should be `INFO`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java [hunk 6]
```diff
@@ -64,18 +63,33 @@ public void execute(SensorContext context) {
...
       LOG.warn("No coverage information will be saved because all LCOV files cannot be found.");
       return;
     }
     saveCoverageFromLcovFiles(context, lcovFiles);
   }
 
+  private static List<File> getLcovFiles(File baseDir, Set<String> reportPaths) {
+    List<File> lcovFiles = new ArrayList<>();
+    for (String reportPath : reportPaths) {
+      LOG.debug("Using pattern '{}' to resolve LCOV files", reportPath);
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java:77]: how about drop `pattern` from this message to not confuse users not using patterns?

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java [hunk 7]
```diff
@@ -285,6 +285,28 @@ public void should_resolve_absolute_path() throws Exception {
     assertThat(context.lineHits(file2Key, 2)).isEqualTo(5);
   }
 
+  @Test
+  public void should_resolve_wildcard_report_paths() throws Exception {
+    settings.setProperty(JavaScriptPlugin.LCOV_REPORT_PATHS, "**/reports/*.lcov");
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java:290]: it's not clear which reports should match

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java (hunk 3 above)
Comment (yassin-kammoun-sonarsource) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java:33]: Agreed

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java (hunk 4 above)
Comment (yassin-kammoun-sonarsource) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/DirectoryScanner.java:43]: Agreed

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java (hunk 2 above)
Comment (yassin-kammoun-sonarsource) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/CoverageSensor.java:78]: Agreed

File: sonar-javascript-plugin/src/test/resources/coverage/wildcard/nested/report.lcov [hunk 8]
```diff
@@ -0,0 +1,12 @@
+TN:
+SF:tests/file1.js
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/resources/coverage/wildcard/nested/report.lcov:2]: let's simplify these 2 reports as much as possible (I think `DA:2,1` is enough)

File: sonar-javascript-plugin/src/test/resources/coverage/wildcard/report.lcov [hunk 9]
```diff
@@ -0,0 +1,29 @@
+TN:
+SF:deep/nested/dir/js/file1.js
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/resources/coverage/wildcard/report.lcov:2]: why having this long path?

File: sonar-javascript-plugin/src/test/resources/coverage/wildcard/report.lcov [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+FNH:0
+FNDA:0,test1
+DA:2,1
+LF:1
+LH:0
+BRF:0
+BRH:0
+end_of_record
+TN:
+SF:file1.js
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/resources/coverage/wildcard/report.lcov:14]: why having another file in the report to prove that it's imported?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**